issue-with-transitions.json

```diff
@@ -0,0 +1,116 @@
+{
+    "$id": "https://example.com/schemas/issue",
+    "$schema": "http://json-schema.org/draft-07/hyper-schema#",
+    "base": "https://example.com/api/",
+    "title": "Issue Tracker",
+    "definitions": {
+        "issue": {
+            "type": "object",
+            "required": ["title", "status"],
+            "properties": {
+                "id": {
+                    "description": "Server-assigned unique identifier; will always be present once assigned during creation",
+                    "type": "integer",
+                    "minimum": 1,
+                    "readOnly": true
+                },
+                "title": {"type": "string"},
+                "status": {
+                    "type": "string",
+                    "enum": ["Open", "Closed"],
+                    "default": "Open"
+                },
+                "description": {"type": "string"}
+            },
+            "links": [
+                {
+                    "description": "",
+                    "rel": "self",
+                    "href": "issue/{id}",
+                    "templateRequired": ["id"],
+                    "targetSchema": {"$ref": "#/definitions/issue"},
+                    "targetHints": {
+                        "allow": ["GET", "PUT"]
+                    },
```

Comment:
Why is a custom extension necessary for a seemingly mundane scenario?

Reply:
What custom extension?  The only extension is `x-dataTransitions` and that's not what this comment is attached to.  All of this target stuff is direct from the draft-07 specification.

Reply:
Sorry again, misfire on the line. Yes, why is x-dataTransitions a necessary extension? Won't this be ignored by basic tooling?

Reply:
Well, there isn't any tooling at the moment (although people are working on that) so that part's momentarily moot- by the time any exists in a few months, this or something analogous could easily be part of the spec.

As for why it's an extension and not *already* in the spec, I have never actually seen a system that really describes data transitions.  By which I mean changes that are all implemented as PUT (or PATCH) but differ only by setting a specific field.  (Note: I have not dug into OpenAPI 3.0's Links Object yet, not sure if it tries to do this).

If you know of any that would work with a PUT/PATCH REST approach, then I would definitely like to see those so I can steal ideas from them :-)

It's a problem that I have thought about some, but since I haven't seen it elsewhere (outside of very RPC-ish systems) I didn't worry about solving it in draft-07.

Reply:
Mostly I just see that sort of thing done in documentation.  The fact that there is a status field with two values "open" and "closed" pretty much tells humans what they need to know, and a client that understands the application (an issue tracker) would be able to figure that out as well.

By "understands the application" I mean that the code is aware that the "status" field is significant and knows the significance of its enumerated values.

Reply:
Although if you are doing an RPC-ish one URI per POST call approach, then you can hardware the field in the request schema (e.g. `"status": {"const": "Open"}` for the "open" call) so that's a way that you could do a data transition of your endpoint is a single-purpose endpoint.

Reply:
> If the issue is "closed", the server includes a link relation "open" that indicates to the user they can re-open the issue. It does not include a "close" link because that transition is not possible.

Ah, I see the problem.  This is an extremely common misunderstanding of the purpose of link relation types.  I looked at link relation types this way myself for several years and ran into many difficulties because of it.

A link relation type is intended to express the relationship between two resources.  The person resource is the *author* of the book resource.  A particular attachment to an issue resource is a *screenshot* of the problem described by the issue.

"open" and "close" are not relationships, they are verbs.  Link relations are *not* verbs.

There are some historical oddities from the Atom specification in the IANA link relation registry that cloud this issue, most notably "edit" which is defined as:

> Refers to a resource that can be used to edit the link's context.

Which is a bit of a tortured definition- what does it mean that it can be "used to edit" the context resource?  But if you look at pretty much [everything else in the registry](https://www.iana.org/assignments/link-relations/link-relations.xhtml) they are not verbs, but relationships.

So, fundamentally, trying to use link relation types for state transitions is incorrect / RPC-ish (whether RPC-ish things should be considered incorrect is a matter of philosophy and use cases- I'm not asserting that it's objectively wrong).

REST is *RE*presentational *S*tate *T*ransfer, which means that you change the state (transition) by changing the representation and transferring it back to the server.

So, Hyper-Schema doesn't do well at expressing "open" and "close" as link relations because they're not, properly speaking, link relations.

Digging any deeper into this critical topic is probably better done on the slack channel.